internal/backend/workloadproxy/reconciler.go

```diff
@@ -6,225 +6,374 @@
+)
 
-// Close closes the load balancer and waits for it to terminate.
+// Reconciler reconciles the load balancers for a list of clusters.
 //
-// It should be called only if it was successfully started, otherwise it blocks indefinitely on the Wait call.
-func (lbSts *lbStatus) Close() error {
-	var lb *loadbalancer.TCP
-
-	lb, lbSts.lb = lbSts.lb, nil
-
-	if lb == nil {
-		return nil
-	}
-
-	if err := lb.Close(); err != nil {
-		return fmt.Errorf("failed to close LB: %w", err)
-	}
-
-	if err := lb.Wait(); err != nil && !errors.Is(err, net.ErrClosed) {
-		return fmt.Errorf("failed to wait for load balancer: %w", err)
-	}
-
-	return nil
-}
-
-// Reconciler reconciles the load balancers for a cluster.
+// Two data strcutures are introduced to esnure internal consistency. Those will be explained below.
+// For now it's important to know two things:
+// 1. The Reconciler will automatically delete the cluster which doesn't have any alias pointing to it.
+// 2. The Reconciler will automatically delete all aliases to the cluster if the cluster was requisted to be removed.
+//
+// For now we use external ttlcache which supports callbacks on eviction. This is used to shut down the active load
+// balancer probe. The expiration time is set to 5 minutes, and reset each time the cluster is accessed directly or via
+// an alias.
+//
+// It starts goroutine to manage the cache, and it will be stopped when the Reconciler is GC'd.
+//
+//nolint:govet
 type Reconciler struct {
-	clusterToAliasToLBStatus map[resource.ID]map[string]*lbStatus
-	aliasToCluster           map[string]resource.ID
-
-	connProvider *memconn.Provider
-	logger       *zap.Logger
-	lbLogger     *zap.Logger
-	logLevel     zapcore.Level
-	mu           sync.Mutex
+	logger    *zap.Logger
+	logLevel  zapcore.Level
+	transport http.RoundTripper
+
+	mu               sync.Mutex
+	shutdown         func()
+	activeProbes     *ttlcache.Cache[resource.ID, remoteList]
+	aliasToCluster   aliasToCluster
+	idleConnections  prometheus.Gauge
+	inUseConnections prometheus.Gauge
 }
 
 // NewReconciler creates a new Reconciler.
 func NewReconciler(logger *zap.Logger, logLevel zapcore.Level) *Reconciler {
-	// use an in-memory transport for the connections to the load balancer
-	provider := &memconn.Provider{}
-
-	provider.MapNetwork("tcp", "memu")
-
 	if logger == nil {
 		logger = zap.NewNop()
 	}
 
-	return &Reconciler{
-		clusterToAliasToLBStatus: map[resource.ID]map[string]*lbStatus{},
-		aliasToCluster:           map[string]resource.ID{},
-		connProvider:             provider,
-		logger:                   logger,
-		lbLogger:                 logger.WithOptions(zap.IncreaseLevel(zapcore.ErrorLevel)),
-		logLevel:                 logLevel,
+	rec := &Reconciler{
+		logger:   logger,
+		logLevel: logLevel,
+		activeProbes: ttlcache.New[resource.ID, remoteList](
+			ttlcache.WithTTL[resource.ID, remoteList](5 * time.Minute),
+		),
+		idleConnections: prometheus.NewGauge(prometheus.GaugeOpts{
```

Comment:
I think for debugging/verification it'd be cool to also have a simple Gauge metric which returns `len(activeProbes)` to make sure we shut them down properly on inactivity.

Reply:
Done!